File: src/main/java/net/hectus/neobb/lore/ItemLoreBuilder.java
package net.hectus.neobb.lore;

import com.marcpg.libpg.lang.Translation;
import com.marcpg.libpg.text.Formatter;
import net.hectus.neobb.turn.Turn;
import net.hectus.neobb.util.Colors;
import net.hectus.neobb.util.Utilities;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.TextDecoration;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Locale;

public abstract class ItemLoreBuilder {
    public static final Component SEPARATOR = Component.text("                    ", Colors.EXTRA);

    protected Turn<?> turn = null;

    public abstract List<Component> build(Locale l);

    public ItemLoreBuilder turn(Turn<?> turn) {
        this.turn = turn;
        return this;
    }

    protected @NotNull Component key(Locale l, String key, String icon) {
        return Component.text(icon + " ").append(Translation.component(l, key)).append(Component.text(": "))
                .color(Colors.NEUTRAL).decoration(TextDecoration.ITALIC, TextDecoration.State.FALSE);
    }

    protected @NotNull List<? extends Component> longText(Locale l, String type) {
        String text = Translation.string(l, type + "." + Utilities.camelToSnake(Utilities.counterFilterName(turn.getClass().getSimpleName())));
        return Formatter.lineWrap(text, 50).stream()
                .map(s -> Component.text("   | " + s, Colors.BLUE)
                        .decoration(TextDecoration.ITALIC, TextDecoration.State.FALSE))
                .toList();
    }
}

File: src/main/kotlin/net/hectus/neobb/lore/ItemLoreBuilder.kt
package net.hectus.neobb.lore

import com.marcpg.libpg.text.Formatter
import net.hectus.neobb.turn.Turn
import net.hectus.neobb.util.Colors
import net.hectus.neobb.util.component
import net.hectus.neobb.util.string
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextDecoration
import java.util.*

abstract class ItemLoreBuilder {
    companion object {
        val SEPARATOR = Component.text("                    ", Colors.EXTRA)
    }

    protected var turn: Turn<*>? = null

    abstract fun build(locale: Locale): List<Component>

    fun buildWithTooltips(locale: Locale): List<Component> {
        val lore = build(locale).toMutableList()
        lore.add(SEPARATOR)
        lore.add(keybindPrefix("key.mouse.left").append(locale.component("item-lore.tooltip.buy")))
        lore.add(keybindPrefix("key.mouse.right").append(locale.component("item-lore.tooltip.buy_3")))
        lore.add(keybindPrefix("key.mouse.left").append(Component.text("+ ")).append(keybindPrefix("key.sneak")).append(locale.component("item-lore.tooltip.buy_max")))
        lore.add(Component.space())
        lore.add(keybindPrefix("key.drop").append(locale.component("item-lore.tooltip.sell")))
        return lore
    }

    fun turn(turn: Turn<*>): ItemLoreBuilder {
        this.turn = turn
        return this
    }

    protected fun key(locale: Locale, key: String, icon: String): Component {
        return Component.text("$icon ").append(locale.component(key)).append(Component.text(": "))
            .color(Colors.NEUTRAL).decoration(TextDecoration.ITALIC, TextDecoration.State.FALSE)
    }

    protected fun longText(l: Locale, type: String): List<Component> {
        val text = l.string("$type." + turn?.namespace())
        return Formatter.lineWrap(text, 50).map { Component.text("   | $it", Colors.BLUE).decoration(TextDecoration.ITALIC, TextDecoration.State.FALSE) }
    }

    protected fun keybindPrefix(key: String): Component {
        return Component.text("[")
            .append(Component.keybind(key))
            .append(Component.text("] "))
    }
}
